### PR TITLE
Add case sensitive directory checks

### PR DESCRIPTION
**Pull request by ilevkivskyi**
This removes one TODO item that used to cause subtle bugs internally.

The test failed on Mac without the fix, I didn't test on Windows, but I think it should be good as well.

### CODE REVIEW

**Comment by msullivan** [comment]
There is a danger with this approach: it checks case sensitivity for the entire path, not just parts of the path that correspond to a python module. This means that if we have something specified in the mypy search path that does not match the canonical casing, we won't be able to find anything in it.

I'm not sure if this is actually a problem? It could be, though. Worth documenting at least.

 Fixing this means changing the API to include a root part that isn't checked, I guess.

**Comment by ilevkivskyi** [comment]
> Fixing this means changing the API to include a root part that isn't checked, I guess.

Unfortunately this causes some code churn, because I need to "thread" this root part (prefix) through callers. Otherwise this looks a simple straightforward fix.

File: mypy/fscache.py [hunk 1]
```diff
@@ -198,9 +201,21 @@ def isfile_case(self, path: str) -> bool:
...
                 names = self.listdir(head)
+                # This allows to check file name case sensitively in
+                # case-insensitive filesystems.
                 res = tail in names and self.isfile(path)
             except OSError:
                 res = False
+
+        # Also check the other path components in case sensitive way.
+        head, dir = os.path.split(head)
+        while res and head and dir and prefix in head:
```
Comment (gvanrossum) [inline mypy/fscache.py]: Why `prefix in head`? That does a substring check and I can think of so many edge cases where this goes wrong (e.g. if prefix is `'/'` or `''`). I would implement a more robust check that checks `prefix` is actually a filesystem prefix of `head`. Or you could remove `prefix` from `head` before you even enter the loop. Or you could use `os.path.relpath()`. Or the caller could be requested to pass the prefix and whatever follows separately (so you'd have pass `os.path.join(prefix, head)` to `listdir()`. Or...

File: test-data/unit/check-modules-case.test [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+[file a/__init__.py]
+[file a/b/__init__.py]
+[file FuNkY_CaSe/other.py]
+x = 1
+
+[file mypy.ini]
+[[mypy]
+mypy_path = tmp/funky_case
+
+[case testPreferPackageOverFileCase]
```
Comment (gvanrossum) [inline test-data/unit/check-modules-case.test]: Looks like the test name is backwards. But don't we also need a test for the reverse? (Where the package has the right case and so is preferred.)

Comment (ilevkivskyi) [reply test-data/unit/check-modules-case.test]: There is a similar test case `testPreferPackageOverFile` in `check-modules.test`. I will add however another one here with `mypy_path` just in case.

File: mypy/fscache.py (hunk 1 above)
Comment (ilevkivskyi) [reply mypy/fscache.py]: Indeed.

File: mypy/modulefinder.py [hunk 3]
```diff
@@ -215,28 +215,31 @@ def _find_module(self, id: str) -> Optional[str]:
         near_misses = []  # Collect near misses for namespace mode (see below).
         for base_dir, verify in candidate_base_dirs:
             base_path = base_dir + seplast  # so e.g. '/usr/lib/python3.4/foo/bar/baz'
+            dir_prefix = base_dir
+            for _ in range(len(components) - 1):
```
Comment (msullivan) [inline mypy/modulefinder.py]: Might be better to strip off `dir_chain` directly with `base_dir[:len(dir_chain)]`?

File: mypy/fscache.py (hunk 1 above)
Comment (ilevkivskyi) [reply mypy/fscache.py]: Actually after playing more with these, it looks like `head.startswith(prefix)` is the only one that behaves correctly, while other options I tried fail in some corner cases. Essentially, if the prefix is `''` or `'/'` we should check the whole path in case sensitive manner.

Also could you please show a tricky test case, so that I can add it to the tests?

File: mypy/modulefinder.py (hunk 3 above)
Comment (ilevkivskyi) [reply mypy/modulefinder.py]: This doesn't work (also it looks like you forgot minus there). I also tried with plus or minus one and some tests fail.

File: mypy/fscache.py [hunk 4]
```diff
@@ -209,7 +209,7 @@ def isfile_case(self, path: str, prefix: str) -> bool:
 
         # Also check the other path components in case sensitive way.
         head, dir = os.path.split(head)
-        while res and head and dir and prefix in head:
+        while res and head and dir and head.startswith(prefix):
```
Comment (gvanrossum) [inline mypy/fscache.py]: This seems good enough. Perhaps you could see if adding `assert head.startswith(prefix)` before line 211 always passes? OTOH perhaps that's risking too much.

I have a faint worry that e.g. `head = '/foo/bar'` and `prefix = '/foo/b'` but that's not possible given how `prefix` is constructed by the caller right?

Comment (ilevkivskyi) [reply mypy/fscache.py]: > Perhaps you could see if adding `assert head.startswith(prefix)` before line 211 always passes? OTOH perhaps that's risking too much.

I added the assert, and all tests still passed. I think however it is better to not put it there, just in case.

> I have a faint worry that e.g. `head = '/foo/bar'` and `prefix = '/foo/b'` but that's not possible given how prefix is constructed by the caller right?

Yes. I will probably just clarify this in the docstring.

**Comment by ilevkivskyi** [comment]
This PR is still labeled as "Changes required", but IIUC the last "LGTM" means it is good to go. So I am merging now.
